### PR TITLE
use try-finally when possible

### PR DESCRIPTION
**Pull request by Tobion**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | I hope |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

Found those with regex `catch \(\\Exception[^\}]+throw \$`

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php [hunk 1]
```diff
@@ -74,13 +74,10 @@ public function load($resource, $type = null)
 
         try {
             $collection = parent::load($resource, $type);
-        } catch (\Exception $e) {
+        } finally {
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php]: @nicolas-grekas I assume this hack also works with finally, i.e. fatal errors are skipping it. I tried to test it as described in #14342. And it had the same result. So this code change didn't change the outcome of the exception page. But it was kinda strange anyway as there was a second error below the real one:

```
FatalErrorException in TraceableEventDispatcher.php line 357:
Error: Cannot use object of type Symfony\Component\EventDispatcher\Debug\WrappedListener as array

in vendor\symfony\symfony\src\Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher.php at line 357
```

File: src/Symfony/Component/Debug/ErrorHandler.php [hunk 2]
```diff
@@ -469,10 +469,8 @@ public function handleError($type, $message, $file, $line, array $context, array
                 $this->isRecursive = true;
                 $this->loggers[$type][0]->log(($type & $level) ? $this->loggers[$type][1] : LogLevel::DEBUG, $message, $e);
                 $this->isRecursive = false;
```
Comment (stof) [inline src/Symfony/Component/Debug/ErrorHandler.php]: this line can be removed now, as it is done in ``finally``

Comment (Tobion) [inline src/Symfony/Component/Debug/ErrorHandler.php]: fixed

**Comment by fabpot** [comment]
Thank you @Tobion.

**Comment by lyrixx** [comment]
@fabpot finally does not work on PHP 5.5.

refs: https://bugs.php.net/bug.php?id=67047

ihmo, symfony should never use finally, except if we bump the requirements on php 5.6

**Comment by nicolas-grekas** [comment]
Well, finally works, but autoloading does not work in a finally block
before 5.6. I reviewed the current patch and it looks safe for 5.5 but we
must have this issue in mind.

**Comment by lyrixx** [comment]
> but we must have this issue in mind.

that's why I think it's saver to recall: "never use it".
Because some people that are not aware of this bug could copy / paste symfony code, or mimic it and hit that bug.

Anyway, I may be too "strict" and may be it's better to be allow it when it's really safe. Don't know ;)

But again, all deciders should be aware of this bug to not merge some PR that trigger autoload.

**Comment by Tobion** [comment]
@lyrixx thanks for pointing out. But as using autoloading in finally is an edge case, it should not prevent us using finally. It's an uncommon case because usually you do something in finally that you already accessed before (a variable, class or whatever to revert state). So autoloading there should not happen.

**Comment by lyrixx** [comment]
@Tobion I totally agree with you ;) In theses case there is no issue. But image this piece of code:

```
try {
  // ..
} finally {
  if (true) {
    throw new NotYetLoadedException()
  }
}
```

;)

**Pull request merged**